django_app/redbox_app/redbox_core/consumers.py

```diff
@@ -255,10 +265,30 @@ async def handle_activity(self, response: dict):
         self.activities.append(RedboxActivityEvent.model_validate(response))
 
     async def handle_documents(self, response: list[Document]):
-        s3_keys = [doc.metadata["file_name"] for doc in response]
-        files = File.objects.filter(original_file__in=s3_keys)
-
+        sources_by_resource_ref = {}
+        for document in response:
+            ref = document.metadata.get("original_resource_ref")
+            docs_from_source = sources_by_resource_ref.get(ref, [])
+            docs_from_source.append(document)
+            sources_by_resource_ref[ref] = docs_from_source
+        handled_sources = set()
+        files = File.objects.filter(original_file__in=sources_by_resource_ref.keys())
```

Comment:
quite complicated, how about?

```python
for ref, sources in sources_by_resource_ref:
    try:
        file = File.objects.filter(original_file=ref)
        await self.send_to_client(
            "source",
            {"original_file_name": file.original_file_name},
        )
        self.citations.append((file, sources_by_resource_ref[file.unique_name]))
    except File.DoesNotExist:
        await self.send_to_client("source", {"url": ref})

        self.citations.extend(
            [(None, doc) for doc in sources]
        )
```

Reply:
Should we add type in the `sources_by_resource_ref` so it contains document's type whether it is a file/uploaded document. 

```
sources_by_resource_ref = {'ref': {'type': 'uploaded_doc', 'documents': [document]}}
```